crates/accelerate/src/convert_2q_block_matrix.rs

```diff
@@ -32,32 +37,75 @@ pub fn blocks_to_matrix(
     py: Python,
     op_list: Vec<(PyReadonlyArray2<Complex64>, SmallVec<[u8; 2]>)>,
 ) -> PyResult<Py<PyArray2<Complex64>>> {
-    let identity = aview2(&ONE_QUBIT_IDENTITY);
-    let input_matrix = op_list[0].0.as_array();
-    let mut matrix: Array2<Complex64> = match op_list[0].1.as_slice() {
-        [0] => kron(&identity, &input_matrix),
-        [1] => kron(&input_matrix, &identity),
+    let identity = aview2(&ONE_QUBIT_IDENTITY).into_faer_complex();
+    let input_matrix = op_list[0].0.as_array().into_faer_complex();
+
+    let mut matrix = match op_list[0].1.as_slice() {
+        [0] => identity.kron(input_matrix),
+        [1] => input_matrix.kron(identity),
         [0, 1] => input_matrix.to_owned(),
-        [1, 0] => change_basis(input_matrix),
-        [] => Array2::eye(4),
+        [1, 0] => change_basis_faer(input_matrix),
+        [] => Mat::<c64>::identity(4, 4),
         _ => unreachable!(),
     };
     for (op_matrix, q_list) in op_list.into_iter().skip(1) {
-        let op_matrix = op_matrix.as_array();
+        let op_matrix = op_matrix.as_array().into_faer_complex();
 
         let result = match q_list.as_slice() {
-            [0] => Some(kron(&identity, &op_matrix)),
-            [1] => Some(kron(&op_matrix, &identity)),
-            [1, 0] => Some(change_basis(op_matrix)),
-            [] => Some(Array2::eye(4)),
+            [0] => Some(identity.kron(op_matrix)),
+            [1] => Some(op_matrix.kron(identity)),
+            [1, 0] => Some(change_basis_faer(op_matrix)),
+            [] => Some(Mat::<c64>::identity(4, 4)),
             _ => None,
         };
-        matrix = match result {
-            Some(result) => result.dot(&matrix),
-            None => op_matrix.dot(&matrix),
+
+        let aux = matrix.clone();
+        match result {
+            Some(x) => {
+                matmul(
+                    matrix.as_mut(),
+                    x.as_ref(),
+                    aux.as_ref(),
+                    None,
+                    c64::new(1., 0.),
+                    Parallelism::None,
+                );
+            }
+            None => {
+                matmul(
+                    matrix.as_mut(),
+                    op_matrix,
+                    aux.as_ref(),
+                    None,
+                    c64::new(1., 0.),
+                    Parallelism::None,
+                );
+            }
```

Comment:
I think all this is just equivalent to:
```rust
let op_matrix = op_matrix.as_array().into_faer_complex();
let expanded = match q_list.as_slice() {
    [0] => Some(identity.kron(op_matrix)),
    [1] => Some(op_matrix.kron(identity)),
    [1, 0] => Some(change_basis_faer(op_matrix)),
    [] => Some(Mat::<c64>::identity(4, 4)),
    _ => None,
};
matrix = expanded.as_ref().map(|x| x.as_ref()).unwrap_or(op_matrix) * matrix;
```

I could be a shade wrong about that if the parallelism effects are different, but otherwise I think this is what we're doing?

Reply:
(see https://github.com/Qiskit/qiskit/pull/12193#discussion_r1579351417)

Reply:
Yeah, this should be equivalent. However, it seems like using the overloaded operator `*` is not as efficient as calling to `matmul`. I created a simple example and these were the results using `cargo bench`:

```
test tests::benchmark04_ndarray_dot          ... bench:    1,430 ns/iter (+/- 115)
test tests::benchmark05_faer_overload_matmul ... bench:    1,419 ns/iter (+/- 410)
test tests::benchmark06_faer_matmul          ... bench:    534 ns/iter (+/- 81) 
```

The tests perform a matrix multiplication in a loop of 5 iterations, using `dot()`, * between two `Mat<Complex64>`, and `matmul()`